dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,42 +52,192 @@ public Pattern getParameterPattern() {
+            throws SQLException {
+        return new DatabaseTable(buildDatabaseObjectName(tableName)).getParams();
     }
 
-    private Map<String, DbParameterAccessor> readIntoParams(
-            String tableOrViewName, String query) throws SQLException {
-        try (PreparedStatement dc = getConnection().prepareStatement(query)) {
-            dc.setString(1, tableOrViewName);
+    private abstract class DatabaseObject {
+        protected DatabaseObjectName name;
+
+        private DatabaseObject(DatabaseObjectName objName) {
+            name = objName;
+        }
+
+        protected DatabaseMetaData getDbMetaData() throws SQLException {
+            return getConnection().getMetaData();
+        }
+
+        abstract protected ResultSet getObjectMetaData() throws SQLException;
+
+        public boolean exists() throws SQLException {
+            try (ResultSet rs = getObjectMetaData()) {
+                return rs.next();
+            }
+        }
+
+        abstract protected ResultSet getColumns() throws SQLException;
 
-            ResultSet rs = dc.executeQuery();
+        public Map<String, DbParameterAccessor> getParams() throws SQLException {
             Map<String, DbParameterAccessor> allParams = new HashMap<String, DbParameterAccessor>();
-            int position = 0;
-            while (rs.next()) {
-                String columnName = rs.getString(1);
-                String dataType = rs.getString(2);
-                DbParameterAccessor dbp = createDbParameterAccessor(
-                        columnName,
-                        Direction.INPUT,
-                        typeMapper.getJDBCSQLTypeForDBType(dataType),
-                        getJavaClass(dataType), position++);
-                allParams.put(NameNormaliser.normaliseName(columnName), dbp);
+            try (ResultSet rs = getColumns()) {
+                while (rs.next()) {
+                    DbParameterAccessor accessor = createColumnAccessor(rs);
+                    allParams.put(normaliseName(accessor.getName()), accessor);
+                }
             }
-            rs.close();
             return allParams;
         }
+
+        protected DbParameterAccessor createColumnAccessor(ResultSet rs)
+                throws SQLException {
+            String paramTypeName = rs.getString("TYPE_NAME");
+            return createDbParameterAccessor(getColumnName(rs), columnDirection(rs),
+                           typeMapper.getJDBCSQLTypeForDBType(paramTypeName),
+                           getJavaClass(paramTypeName), rs.getInt("ORDINAL_POSITION") - 1);
+        }
+
+        protected String getColumnName(ResultSet rs) throws SQLException {
+            return defaultIfNull(rs.getString("COLUMN_NAME"), "");
+        }
+
+        protected Direction columnDirection(ResultSet rs) throws SQLException {
+            Direction columnDir = getColumnDirection(rs);
+            if (columnDir == null) {
```

Comment:
This method isn't defined as throwing an NPE.

https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#requireNonNull(T,%20java.lang.String)

So I'm not required to `throws NullPointerException`.

We still have `SQLExceptions` from `getColumnDirection` because the the `rs` field access.